### ISSUE TITLE
Update the indices for the fieldSplit preconditioner 

### ISSUE THREAD

**Issue by fdkong** [idaholab/moose#7542]
### Description of the enhancement or error report
We used to update the indices for the fieldSplit preconditioner for every nonlinear iteration. It is expensive to do so. For most cases, it is necessary to update the indices for every timestep only.

**Comment by friedmud** [comment]
This should be an option!  Or should be tied to the options you use in the geometric search system!

Just because it's working with our current tests / examples doesn't mean that's always going to work.  It's completely possible to need these updated every nonlinear iteration (in fact, it's possible to need them updated every _linear_ iteration!).

In cases where the geometric search system is not impacting the splits (i.e. no `ContactSplit` type of thing is going on) then these indices actually should only get updated when the mesh changes...

**Comment by fdkong** [comment]
@friedmud, I agree with you. That is why I said that I just temporally comment out a line of code. We do not have a generic method to determine when the indices should be updated. I was thinking that we could use the function, MatGetNonzeroState, from PETSc. The indices will be updated only when the nonzero structure of Mat has been changed. However, I have to check how we deal with Mat in MOOSE, especially, the preallocation part. 

Do you have any idea?

**Comment by friedmud** [comment]
I think that for right now you could simply let the user choose when to update them using the normal `execute_on` options of `initial`, `timestep_begin`, `residual`, `nonlinear`, etc.

As for doing it automatically... you would have to watch `FEProblem::augmentSparsity()`...

**Comment by fdkong** [comment]
@friedmud, Thanks. It is a good idea. But not all options make sense. For the fieldSplit preconditioner, I think there could be three options "nonlinear", "timestep" and "auto". "auto" means automatically update the indices according to the structure change of the matrix.

**Comment by friedmud** [comment]
Why not `initial`?  I would definitely want that option sometimes... it would just mean to compute it once at the beginning and never change it.

Is `residual` not a possibility?  Because it _is_ a possibility in our geometric search system (it gets updated during the residual evaluation - so dof couplings _can_ change).

Also: I would want both `timestep_end` and `timestep_begin`.

**Comment by fdkong** [comment]
> Why not initial? I would definitely want that option sometimes... it would just mean to compute it 
> once at the beginning and never change it.

Yes, this makes sense for me.
After each time step, the nonlinear solver object, snes, will be destroyed in Libmesh, and inner objects such as pc, ksp and dm are all destroyed. To keep using the same indices for the entire simulation period, we have to somehow cache the dm before SNES tries to destroy it. Is it OK for you if I try to cache a dm in the _fe_problem? I am trying to avoid messing up the PETSc code with others. 

>Is residual not a possibility? Because it is a possibility in our geometric search system (it gets >updated during the residual evaluation - so dof couplings can change).

I agree that the dof couplings could change during every residual evaluation. But the preconditioning matrix will be updated only for each nonlinear iteration not for each linear iteration. The preconditioner will NOT have any update until next nonlinear iteration, even though we update the indices to reflect the changes of the dof couplings. Therefore, I do NOT think this option will be going to do anything.


> Also: I would want both timestep_end and timestep_begin.

"timestep_begin" makes sense for me. But, it is not necessary to have the option, "timestep_end". After each time step, we already have the solution for that time step,  the preconditioning job is done and we do not have more nonlinear systems to solve except stepping into the next time step. For the next time step, the option  "timestep_begin" will handle the update.  

Pinging @friedmud.

**Comment by friedmud** [comment]
It sounds to me like this should be resolved at the libMesh level.  How about we add an option to the libMesh solver objects to _NOT_ destroy the PETSc objects.

I actually thought there was some discussion about doing just that a little while ago on the libMesh mailing list... but I can't find it now...

@roystgnr do you remember that?

**Comment by roystgnr** [comment]
I recall the discussion but can't find it either.  I'd definitely support making that an option, probably even making it the default option after a brief period.
